### ISSUE TITLE
Broken debug info with 1.62.0

### ISSUE THREAD

**Issue by glandium** [rust-lang/rust#98746]
Building Firefox with rustc 1.62.0 on Linux and running it under valgrind makes valgrind emit the following error:
```
==31382== Valgrind: debuginfo reader: ensure_valid failed:
==31382== Valgrind:   during call to ML_(img_get)
==31382== Valgrind:   request for range [3275670898, +4) exceeds
==31382== Valgrind:   valid image size of 1880027600 for image:
==31382== Valgrind:   "/builds/worker/workspace/obj-build/toolkit/library/build/libxul.so"
==31382==
==31382== Valgrind: debuginfo reader: Possibly corrupted debuginfo file.
==31382== Valgrind: I can't recover.  Giving up.  Sorry.
```
As of filing, I don't have more information to share yet, but I thought I'd already file it before starting to investigate it next week.

**Comment by aordano** [comment]
I am unfamiliar with Valgrind, but i updated a couple hours ago and i can't get LLDB to display local variables while debugging. Maybe related? It worked perfectly fine before the updated and i didn't change anything else.

Tried common fixes, resetting, etc.

**Comment by glandium** [comment]
This seems to have started somewhere between nightly-2022-04-28 and nightly-2022-04-29... and there's an LLVM update in the range. I'm going to bisect further with the bors builds.

**Comment by glandium** [comment]
Yep, started on 1388b38c52d1ca9fbc80bf42fa007504fb0b1b41.

**Comment by glandium** [comment]
So... if I build 1388b38 locally, it doesn't reproduce...

**Comment by glandium** [comment]
The CI change in 278fdf6c828f18d2db1b3b5d026214ce2d7dadb3 had the unintended side effect of shipping libcompiler-builtins with DWARF5 debuginfo, which is not well supported.

**Comment by glandium** [comment]
For the record, libprofiler_builtins, and librustc-stable_rt.*san.a also have DWARF5 debuginfo.

**Comment by glandium** [comment]
For the record (bis), my local builds of 1.62 were unaffected despite being compiled with clang 14 because they don't contain debuginfo in those libs at all.

**Comment by stevenengler** [comment]
This also seems to prevent me from debugging binaries with rust code in gdb. For example with a C project that statically links a rust library on Ubuntu 20.04 with rustc/cargo 1.62.0 and 1.63.0:

```
$ gdb --args /my/binary
Reading symbols from /my/binary...
Dwarf Error: DW_FORM_strx1 found in non-DWO CU [in module /my/binary]
(No debugging symbols found in /my/binary)
```

Then gdb isn't able to show debugging symbols for most of the code.

An llvm thread says this is a bug with DWARF 5 and gdb: https://discourse.llvm.org/t/gdb-10-1-cant-read-clangs-dwarf-v5/6035

It's only a single part of the binary that has DWARF version 5:

```
$ llvm-dwarfdump -r 1 /my/binary
.debug_info contents:
0x00000000: Compile Unit: length = 0x000012fb version = 0x0004 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x000012ff)

[... a bunch of other compilation units with version = 0x0004]

0x010a6be3: Compile Unit: length = 0x0000008f version = 0x0005 unit_type = DW_UT_compile abbr_offset = 0x83eff addr_size = 0x08 (next unit at 0x010a6c76)

0x010a6bef: DW_TAG_compile_unit
              DW_AT_producer	("clang version 14.0.1")
              DW_AT_language	(DW_LANG_C99)
              DW_AT_name	("./lib/builtins/popcountdi2.c")
              DW_AT_str_offsets_base	(0x00000008)
              DW_AT_stmt_list	(0x0065655b)
              DW_AT_comp_dir	("/cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.71")
              DW_AT_low_pc	(0x000000000147eb10)
              DW_AT_high_pc	(0x000000000147eb72)
              DW_AT_addr_base	(0x00000008)
              DW_AT_loclists_base	(0x0000000c)
```

If I build my project with rustc/cargo 1.61.0, then I'm able to debug it in gdb and the above gdb error goes away.

This underlying issue seems to be a bug with gdb < 10.1 (such as version 9.2 on ubuntu 20.04), but just wanted to mention a workflow that using DWARF 5 in compiler_builtins seems to break.

**Comment by nnethercote** [comment]
FWIW, the next release of Valgrind (3.20) will have better support for DWARF 5 and should be out in the next week or two.

**Comment by pnkfelix** [comment]
@rustbot label: I-compiler-nominated

nominating for discussion at T-compiler meeting. 

1. I suspect this is a real bug that we need to address a bit more aggressively than we have been. (Who knows how many rust users are impacted, and what percentage of them will not be prepared to upgrade to the next valgrind release.)
2. It might be a sign of a process issue that we didn't flag it as something that needs prioritization. (I blame myself here as well; I looked at this issue 10 days ago, but didn't make the connection that it wasn't tagged T-compiler and wasn't prioritized.)

**Comment by nnethercote** [comment]
Valgrind 3.20 has been released.

**Comment by glandium** [comment]
Note that valgrind is not the only tool that has problems with DWARF 5.

**Comment by glandium** [comment]
Also note there's been a PR on cc-rs for this for a while.

**Comment by apiraino** [comment]
@glandium I think you're referring to this one PR on cc-rs: https://github.com/rust-lang/cc-rs/pull/694

(linking it here for reference)

**Comment by apiraino** [comment]
Assigning priority as suggestd in T-compiler meeting ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-11-10/near/309019592)).

@rustbot label -I-prioritize +P-high -I-compiler-nominated

**Comment by RalfJung** [comment]
Sadly Debian still doesn't have Valgrind 3.20, even it 'unstable', so this issue makes it quite hard to e.g. profile rustc performance issues on Debian.

**Comment by glandium** [comment]
This should actually be fixed as of [5645d10](https://github.com/compiler-errors/rust/commit/5645d10f93c541f88da21163a71b326aa6f3f563)

**Comment by RalfJung** [comment]
I'm still seeing these valgrind errors on current rustc (as in, when I am trying to profile rustc itself).

**Comment by glandium** [comment]
Oh indeed, while libcompiler_builtins is DWARF4, the rustc binary itself is DWARF5...

**Comment by glandium** [comment]
... which is weird because the rust compiler itself defaults to emitting DWARF 4 at most.

**Comment by Enselic** [comment]
Triage: Since quite a lot of time has passed without a fix it seems like this was not a super big issue in practice. And the longer we wait, the less important it becomes to fix, since DWARF5 support in tools becomes more and more widely available. Eventually, fixing is not needed at all (arguably we are already there). I suggest we close this issue as "won't fix".

**Comment by glandium** [comment]
Another way to look at it is that rustc is not being explicit about what it wants, and that's likely to cause similar problems in the future when a LLVM update makes it use DWARF6. Half the problem was fixed by having cc-rs hard code something that kind of matches what rustc supposedly does, but apparently, that wasn't enough.

So yes, maybe the problem is not so much of a problem today with DWARF5, but it's also not future-proof.
